File: app/realtime.py
"""
Real-time Communication Module for Enhanced Workflow
Handles WebSocket connections, notifications, and live updates
"""

from flask import request, session
from flask_socketio import SocketIO, emit, join_room, leave_room
import json
import redis
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

# Initialize Redis for message handling
redis_client = redis.Redis(host='localhost', port=6379, db=0, decode_responses=True)

class RealtimeManager:
    """Manages real-time communication and notifications"""
    
    def __init__(self, socketio):
        self.socketio = socketio
        self.setup_handlers()
    
    def setup_handlers(self):
        """Setup WebSocket event handlers"""
        
        @self.socketio.on('connect')
        def handle_connect():
            """Handle client connection"""
            user_id = session.get('user_id')
            if user_id:
                join_room(f"user_{user_id}")
                logger.info(f"User {user_id} connected to WebSocket")
                emit('connection_status', {'status': 'connected'})
            else:
                emit('connection_status', {'status': 'unauthorized'})
                return False
        
        @self.socketio.on('disconnect')
        def handle_disconnect():
            """Handle client disconnection"""
            user_id = session.get('user_id')
            if user_id:
                leave_room(f"user_{user_id}")
                logger.info(f"User {user_id} disconnected from WebSocket")
        
        @self.socketio.on('join_application')
        def handle_join_application(data):
            """Join application-specific room for real-time updates"""
            application_id = data.get('application_id')
            user_id = session.get('user_id')
            
            if application_id and user_id:
                room_name = f"app_{application_id}"
                join_room(room_name)
                emit('joined_application', {'application_id': application_id})
                logger.info(f"User {user_id} joined application room {room_name}")
        
        @self.socketio.on('leave_application')
        def handle_leave_application(data):
            """Leave application-specific room"""
            application_id = data.get('application_id')
            user_id = session.get('user_id')
            
            if application_id and user_id:
                room_name = f"app_{application_id}"
                leave_room(room_name)
                emit('left_application', {'application_id': application_id})
                logger.info(f"User {user_id} left application room {room_name}")
        
        @self.socketio.on('send_message')
        def handle_send_message(data):
            """Handle real-time messaging between users and analysts"""
            message = data.get('message')
            application_id = data.get('application_id')
            user_id = session.get('user_id')
            
            if not all([message, application_id, user_id]):
                emit('error', {'message': 'Missing required fields'})
                return
            
            # Store message in database
            message_data = {
                'id': f"msg_{datetime.now().timestamp()}",
                'application_id': application_id,
                'user_id': user_id,
                'message': message,
                'timestamp': datetime.now().isoformat()
            }
            
            # Emit to all users in the application room
            self.socketio.emit('new_message', message_data, room=f"app_{application_id}")
            
            # Store in Redis for persistence
            redis_client.lpush(f"messages:app_{application_id}", json.dumps(message_data))
            redis_client.expire(f"messages:app_{application_id}", 86400)  # 24 hours
        
        @self.socketio.on('request_clarification')
        def handle_request_clarification(data):
            """Handle clarification requests from analysts"""
            application_id = data.get('application_id')
            question_id = data.get('question_id')
            message = data.get('message')
            analyst_id = session.get('user_id')
            
            if not all([application_id, question_id, message, analyst_id]):
                emit('error', {'message': 'Missing required fields'})
                return
            
            # Create clarification request
            clarification_data = {
                'id': f"clar_{datetime.now().timestamp()}",
                'application_id': application_id,
                'question_id': question_id,
                'analyst_id': analyst_id,
                'message': message,
                'timestamp': datetime.now().isoformat()
            }
            
            # Notify user
            self.socketio.emit('clarification_request', clarification_data, 
                             room=f"app_{application_id}")
            
            # Store in Redis
            redis_client.lpush(f"clarifications:app_{application_id}", json.dumps(clarification_data))
    
    def notify_status_change(self, application_id, new_status, user_id, message=None):
        """Notify users of status changes"""
        notification_data = {
            'type': 'status_change',
            'application_id': application_id,
            'status': new_status,
            'message': message or f"Status changed to {new_status}",
            'timestamp': datetime.now().isoformat()
        }
        
        # Emit to application room
        self.socketio.emit('status_update', notification_data, room=f"app_{application_id}")
        
        # Emit to specific user
        self.socketio.emit('status_update', notification_data, room=f"user_{user_id}")
    
    def notify_progress_update(self, application_id, progress_percentage, milestone=None):
        """Notify users of progress updates"""
        progress_data = {
            'type': 'progress_update',
            'application_id': application_id,
            'progress': progress_percentage,
            'milestone': milestone,
            'timestamp': datetime.now().isoformat()
        }
        
        # Emit to application room
        self.socketio.emit('progress_update', progress_data, room=f"app_{application_id}")
    
    def notify_assignment(self, application_id, analyst_id, assignment_type='automatic'):
        """Notify analyst of new assignment"""
        assignment_data = {
            'type': 'assignment',
            'application_id': application_id,
            'assignment_type': assignment_type,
            'timestamp': datetime.now().isoformat()
        }
        
        # Emit to analyst
        self.socketio.emit('new_assignment', assignment_data, room=f"user_{analyst_id}")
    
    def notify_clarification_response(self, application_id, analyst_id, response):
        """Notify analyst of clarification response"""
        response_data = {
            'type': 'clarification_response',
            'application_id': application_id,
            'response': response,
            'timestamp': datetime.now().isoformat()
        }
        
        # Emit to analyst
        self.socketio.emit('clarification_response', response_data, room=f"user_{analyst_id}")
    
    def broadcast_system_notification(self, message, notification_type='info', target_roles=None):
        """Broadcast system-wide notifications"""
        notification_data = {
            'type': 'system_notification',
            'message': message,
            'notification_type': notification_type,
            'timestamp': datetime.now().isoformat()
        }
        
        if target_roles:
            for role in target_roles:
                self.socketio.emit('system_notification', notification_data, room=f"role_{role}")
        else:
            self.socketio.emit('system_notification', notification_data)
    
    def get_application_messages(self, application_id, limit=50):
        """Get recent messages for an application"""
        messages = redis_client.lrange(f"messages:app_{application_id}", 0, limit-1)
        return [json.loads(msg) for msg in messages]
    
    def get_application_clarifications(self, application_id, limit=20):
        """Get recent clarifications for an application"""
        clarifications = redis_client.lrange(f"clarifications:app_{application_id}", 0, limit-1)
        return [json.loads(clar) for clar in clarifications]

# Global realtime manager instance
realtime_manager = None

def init_realtime(socketio):
    """Initialize real-time communication"""
    global realtime_manager
    realtime_manager = RealtimeManager(socketio)
    return realtime_manager

def get_realtime_manager():
    """Get the global realtime manager instance"""
    return realtime_manager

File: static/js/realtime.js
/**
 * Real-time Communication JavaScript
 * Handles WebSocket connections and real-time updates
 */

class RealtimeManager {
    constructor() {
        this.socket = null;
        this.connected = false;
        this.currentApplicationId = null;
        this.messageHandlers = new Map();
        this.setupEventListeners();
    }

    init() {
        // Initialize Socket.IO connection
        this.socket = io();
        
        this.socket.on('connect', () => {
            this.connected = true;
            this.updateConnectionStatus(true);
            console.log('Connected to real-time server');
        });

        this.socket.on('disconnect', () => {
            this.connected = false;
            this.updateConnectionStatus(false);
            console.log('Disconnected from real-time server');
        });

        this.socket.on('connection_status', (data) => {
            if (data.status === 'connected') {
                this.connected = true;
                this.updateConnectionStatus(true);
            } else {
                this.connected = false;
                this.updateConnectionStatus(false);
            }
        });

        // Set up message handlers
        this.setupMessageHandlers();
    }

    setupMessageHandlers() {
        // Status updates
        this.socket.on('status_update', (data) => {
            this.handleStatusUpdate(data);
        });

        // Progress updates
        this.socket.on('progress_update', (data) => {
            this.handleProgressUpdate(data);
        });

        // New messages
        this.socket.on('new_message', (data) => {
            this.handleNewMessage(data);
        });

        // Clarification requests
        this.socket.on('clarification_request', (data) => {
            this.handleClarificationRequest(data);
        });

        // Clarification responses
        this.socket.on('clarification_response', (data) => {
            this.handleClarificationResponse(data);
        });

        // New assignments
        this.socket.on('new_assignment', (data) => {
            this.handleNewAssignment(data);
        });

        // System notifications
        this.socket.on('system_notification', (data) => {
            this.handleSystemNotification(data);
        });
    }

    setupEventListeners() {
        // Auto-join application room when page loads
        document.addEventListener('DOMContentLoaded', () => {
            const appId = this.getCurrentApplicationId();
            if (appId) {
                this.joinApplication(appId);
            }
        });

        // Handle page visibility changes
        document.addEventListener('visibilitychange', () => {
            if (document.hidden) {
                this.pauseUpdates();
            } else {
                this.resumeUpdates();
            }
        });
    }

    getCurrentApplicationId() {
        // Try to get application ID from URL or data attributes
        const path = window.location.pathname;
        const appIdMatch = path.match(/\/applications\/([^\/]+)/);
        if (appIdMatch) {
            return appIdMatch[1];
        }

        // Check for data attribute
        const appElement = document.querySelector('[data-application-id]');
        if (appElement) {
            return appElement.getAttribute('data-application-id');
        }

        return null;
    }

    joinApplication(applicationId) {
        if (this.socket && this.connected) {
            this.currentApplicationId = applicationId;
            this.socket.emit('join_application', { application_id: applicationId });
        }
    }

    leaveApplication(applicationId) {
        if (this.socket && this.connected) {
            this.socket.emit('leave_application', { application_id: applicationId });
            this.currentApplicationId = null;
        }
    }

    sendMessage(applicationId, message) {
        if (this.socket && this.connected) {
            this.socket.emit('send_message', {
                application_id: applicationId,
                message: message
            });
        }
    }

    requestClarification(applicationId, questionId, message) {
        if (this.socket && this.connected) {
            this.socket.emit('request_clarification', {
                application_id: applicationId,
                question_id: questionId,
                message: message
            });
        }
    }

    // Message handlers
    handleStatusUpdate(data) {
        console.log('Status update received:', data);
        
        // Update status in UI
        const statusElement = document.querySelector(`[data-application-id="${data.application_id}"] .status-badge`);
        if (statusElement) {
            statusElement.textContent = data.status;
            statusElement.className = `badge bg-${this.getStatusColor(data.status)}`;
        }

        // Show notification
        this.showNotification('Status Update', data.message, 'info');
        
        // Trigger custom event
        this.triggerCustomEvent('statusUpdate', data);
    }

    handleProgressUpdate(data) {
        console.log('Progress update received:', data);
        
        // Update progress bar
        const progressElement = document.querySelector(`[data-application-id="${data.application_id}"] .progress-bar`);
        if (progressElement) {
            progressElement.style.width = `${data.progress}%`;
            progressElement.setAttribute('aria-valuenow', data.progress);
        }

        // Update progress text
        const progressText = document.querySelector(`[data-application-id="${data.application_id}"] .progress-text`);
        if (progressText) {
            progressText.textContent = `${data.progress}% Complete`;
        }

        // Show milestone notification
        if (data.milestone) {
            this.showNotification('Progress Update', `Milestone: ${data.milestone}`, 'success');
        }

        // Trigger custom event
        this.triggerCustomEvent('progressUpdate', data);
    }

    handleNewMessage(data) {
        console.log('New message received:', data);
        
        // Add message to chat
        this.addMessageToChat(data);
        
        // Show notification
        this.showNotification('New Message', 'You have a new message', 'info');
        
        // Trigger custom event
        this.triggerCustomEvent('newMessage', data);
    }

    handleClarificationRequest(data) {
        console.log('Clarification request received:', data);
        
        // Show clarification modal
        this.showClarificationModal(data);
        
        // Show notification
        this.showNotification('Clarification Request', 'Analyst needs clarification', 'warning');
        
        // Trigger custom event
        this.triggerCustomEvent('clarificationRequest', data);
    }

    handleClarificationResponse(data) {
        console.log('Clarification response received:', data);
        
        // Update UI to show response received
        this.updateClarificationStatus(data.application_id, 'responded');
        
        // Show notification
        this.showNotification('Clarification Response', 'User provided clarification', 'success');
        
        // Trigger custom event
        this.triggerCustomEvent('clarificationResponse', data);
    }

    handleNewAssignment(data) {
        console.log('New assignment received:', data);
        
        // Update assignment UI
        this.updateAssignmentStatus(data.application_id, 'assigned');
        
        // Show notification
        this.showNotification('New Assignment', 'You have been assigned a new review', 'info');
        
        // Trigger custom event
        this.triggerCustomEvent('newAssignment', data);
    }

    handleSystemNotification(data) {
        console.log('System notification received:', data);
        
        // Show system notification
        this.showNotification('System Notification', data.message, data.notification_type);
        
        // Trigger custom event
        this.triggerCustomEvent('systemNotification', data);
    }

    // UI Helper methods
    updateConnectionStatus(connected) {
        const statusElement = document.getElementById('connectionStatus');
        if (statusElement) {
            if (connected) {
                statusElement.innerHTML = '<i class="fas fa-circle"></i> Connected';
                statusElement.className = 'badge bg-success';
            } else {
                statusElement.innerHTML = '<i class="fas fa-circle"></i> Disconnected';
                statusElement.className = 'badge bg-danger';
            }
        }
    }

    getStatusColor(status) {
        const statusColors = {
            'draft': 'secondary',
            'submitted': 'warning',
            'in_review': 'primary',
            'pending_clarification': 'warning',
            'completed': 'success',
            'rejected': 'danger'
        };
        return statusColors[status] || 'secondary';
    }

    showNotification(title, message, type = 'info') {
        // Create notification element
        const notification = document.createElement('div');
        notification.className = `alert alert-${type} alert-dismissible fade show position-fixed`;
        notification.style.top = '20px';
        notification.style.right = '20px';
        notification.style.zIndex = '9999';
        notification.style.minWidth = '300px';
        
        notification.innerHTML = `
            <strong>${title}</strong><br>
            ${message}
            <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
        `;
        
        document.body.appendChild(notification);
        
        // Auto-remove after 5 seconds
        setTimeout(() => {
            if (notification.parentNode) {
                notification.remove();
            }
        }, 5000);
    }

    addMessageToChat(messageData) {
        const chatContainer = document.getElementById('chatMessages');
        if (chatContainer) {
            const messageElement = document.createElement('div');
            messageElement.className = 'message-item mb-2 p-2 border rounded';
            messageElement.innerHTML = `
                <div class="d-flex justify-content-between">
                    <strong>User ${messageData.user_id}</strong>
                    <small class="text-muted">${new Date(messageData.timestamp).toLocaleTimeString()}</small>
                </div>
                <div>${messageData.message}</div>
            `;
            
            chatContainer.appendChild(messageElement);
            chatContainer.scrollTop = chatContainer.scrollHeight;
        }
    }

    showClarificationModal(data) {
        // Update modal content
        document.getElementById('clarificationRequestText').textContent = data.message;
        document.getElementById('clarificationApplicationId').value = data.application_id;
        document.getElementById('clarificationQuestionId').value = data.question_id;
        
        // Show modal
        const modal = new bootstrap.Modal(document.getElementById('clarificationModal'));
        modal.show();
    }

    updateClarificationStatus(applicationId, status) {
        const statusElement = document.querySelector(`[data-application-id="${applicationId}"] .clarification-status`);
        if (statusElement) {
            statusElement.textContent = status;
            statusElement.className = `badge bg-${status === 'responded' ? 'success' : 'warning'}`;
        }
    }

    updateAssignmentStatus(applicationId, status) {
        const statusElement = document.querySelector(`[data-application-id="${applicationId}"] .assignment-status`);
        if (statusElement) {
            statusElement.textContent = status;
            statusElement.className = `badge bg-${status === 'assigned' ? 'success' : 'secondary'}`;
        }
    }

    pauseUpdates() {
        // Pause real-time updates when page is not visible
        console.log('Pausing real-time updates');
    }

    resumeUpdates() {
        // Resume real-time updates when page becomes visible
        console.log('Resuming real-time updates');
    }

    triggerCustomEvent(eventName, data) {
        // Trigger custom events for other parts of the application
        const event = new CustomEvent(eventName, { detail: data });
        document.dispatchEvent(event);
    }

    // Public API methods
    on(eventName, handler) {
        if (!this.messageHandlers.has(eventName)) {
            this.messageHandlers.set(eventName, []);
        }
        this.messageHandlers.get(eventName).push(handler);
    }

    off(eventName, handler) {
        if (this.messageHandlers.has(eventName)) {
            const handlers = this.messageHandlers.get(eventName);
            const index = handlers.indexOf(handler);
            if (index > -1) {
                handlers.splice(index, 1);
            }
        }
    }
}

// Global realtime manager instance
window.realtimeManager = new RealtimeManager();

// Initialize when DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
    window.realtimeManager.init();
});

// Export for module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = RealtimeManager;
}
